Blog Post: Advancements in eCash (#481)
OpenSats/website: data/blog/advancements-in-ecash.mdx

>organizations is being made each time any government or business decides to
>automate another set of transactions. In one direction lies unprecedented
>scrutiny and control of people's lives, in the other, secure parity between
>individuals and organizations. The shape of society in the next century may
>depend on which approach predominates.
>
><cite>—[David
>Chaum](https://groups.csail.mit.edu/mac/classes/6.805/articles/money/chaum-electronic-privacy.html)</cite>

Over the past year, OpenSats has given long-term and targeted support to ecash
developers and projects. This led to contributors proposing new protocol
standards, improving reference implementations, enhancing wallet usability, and
adding entry points for merchants and services. Several new standards now span
multiple projects, signaling adoption, interoperability, and the growth of
bitcoin-based ecash infrastructure.

The nine initiatives highlighted in this impact report are:

- [Cashu](#cashu)
- [Cashu Dev Kit](#cashu-dev-kit)
- [Nutshell](#nutshell)
- [Minibits Wallet](#minibits-wallet)
- [Nutstash](#nutstash)
- [Npub.cash](#npubcash)
- [Hashpool](#hashpool)
- [Fedimint](#fedimint)
- [OpenCash](#opencash)

> The cumulative advancements of these projects  mark a shift from experimentation

Reviewer: Sounds awkward to me. Maybe this instead -- feel free to ignore.

```suggestion
The cumulative advancement of these projects marks a shift from experimentation
```

Hashpool also explored how ecash payouts could be extended through Lightning and
payment requests, with an [active pull
request](https://github.com/vnprc/hashpool/pull/62) open for Lightning
integration. These additions would allow miners to instantly redeem their tokens
for bitcoin, thereby broadening the utility of ecash-based rewards.

A related project called
[Axepool](https://hashpool.dev/dev-logs/dev-log-2025-04-05/#axepool) applies the
same design to small-scale miners, particularly those running
[Bitaxe](/blog/bitcoin-grants-feb-2024#the-bitaxe) hardware. By issuing
[eHash](https://hashpool.dev/articles/what-is-hashpool/#ehash) (ecash tokens
backed by proof of work) directly for contributed hashpower, Axepool removes
account requirements and minimum withdrawal limits, making it easier for
everyday miners to receive frequent payouts. Together, Hashpool and Axepool
demonstrate how ecash can replace traditional pool accounting with direct,
accountless mining rewards that are private and transferable.

### Fedimint

Fedimint is an ecash protocol based on a federated Chaumian e-cash framework,
designed to let communities pool custody and issue private bearer tokens. With

> OpenSats' support, contributors made progress delivering software and tooling.

Reviewer: This is the only OpenSats support mention in this post that doesn't contain a link.

Redis-backed caching, [Keycloak-based
authentication](https://github.com/cashubtc/nutshell/pull/675), and [expanded
quote responses](https://github.com/cashubtc/nuts/pull/235) that ease wallet
integration in stateless environments.

Alongside protocol changes, Calle developed tools to improve reliability and
adoption. The [Cashu Auditor](https://github.com/callebtc/cashu-auditor)
automates test payments between mints and exposes uptime through an API, giving
wallet UIs visibility into mint health.
[NFChat](https://github.com/callebtc/nfchat) demonstrated NFC-based transfers
between Android wallets, and the [Cashu
Decoder](https://github.com/callebtc/cashu-decoder) made it easier for
developers to inspect and debug tokens.

>Cashu has grown into a true pillar of the Bitcoin open-source movement, driving
>innovations once thought impossible. The free and open-source project has
>brought countless new developers into Bitcoin. Today, Cashu powers Bitcoin
>communities worldwide, and is already at the heart of multiple companies’
>products. This momentum is only possible thanks to our incredible grassroots
>community and the freedom-driven contributors who make Cashu what it is today.
>

> ><cite>---Calle</cite>

Author: ```suggestion
><cite>—Calle</cite>
```

[m1sterc001guy](/blog/m1sterc001guy-receives-lts-grant), introduced
[LNv2](https://github.com/fedimint/fedimint/pull/7240), a new payment system
that fixes malleability and simplifies invoices. Operators also gained a
self-contained [LDK mode](https://github.com/fedimint/fedimint/pull/5540) that
eliminates the need for an external Lightning node, along with improved fee
handling, authenticated endpoints, and simple [12-word
recovery](https://github.com/fedimint/fedimint/pull/5908) for both ecash and
on-chain funds.

Networking and deployment were improved by Harbor and core contributors, led by
[oleonardolima](https://github.com/oleonardolima). `Tor`
[support](https://github.com/fedimint/fedimint/pull/5443) was added for private
connectivity, while `Iroh` [peer-to-peer
networking](https://github.com/fedimint/fedimint/pull/7042) removed the need for
domain names or port forwarding. Packaging advances delivered [`Docker
Compose`](https://github.com/fedimint/fedimint/pull/7042) files and a
[`Start9`](https://github.com/fedimint/fedimint/pull/7117) service, making
federations easier to launch and manage.

Wallet safety and upgrades were the focus of core releases coordinated by

> [Bradley Stachurski](https://github.com/bradleystachurski). A [double-spend
> vulnerability](https://github.com/fedimint/fedimint/pull/6065) was fixed, while

Author: ```suggestion
[Bradley Stachurski](https://github.com/bradleystachurski), while
```

The mint role was [refactored](https://github.com/vnprc/hashpool/commit/bc50f53535dddc4eea0bb0569bf1768ada862d9b) into a proper Sv2 role using all new [extension](https://stratumprotocol.org/specification/03-Protocol-Overview/#34-protocol-extensions) message types, and the hashpool repo was upgraded with [Nix
Flakes](https://nix.dev/concepts/flakes.html), Bitcoin
[regtest](https://developer.bitcoin.org/examples/testing.html) setups, and
[command-line utilities](https://github.com/vnprc/hashpool/commit/28b682df9023babbce60be34ef0e612c58f3e65d) for
testing wallet balances and redemption. Together, these changes make Hashpool
easier to test and operate under real-world mining conditions.

Hashpool developer vnprc also vibed with nostr builders at [Sovereign Engineering](https://sovereignengineering.io/) Cohort 5 resulting in a [published rust crate](https://crates.io/crates/bitcoin-nostr-relay) for transmitting bitcoin transactions via nostr,

> and experiments in ehash payment forwarding via nostr, nix build systems, and a slick new [user interface](https://x.com/btchashpool/status/1970850087719809311). These tools will

Reviewer 2: don't use a twitter link to talk about nostr for goodness sake! njump is struggling but this one seems stable. alternatives include: njump.me, nostr.at, and nostr.ae

```suggestion
and experiments in ehash payment forwarding via nostr, nix build systems, and a slick new [user interface](https://nostr.eu/nevent1qqsw2q6nsactqg5c7t3r0se3l276r0xd0agjae6xlv38nzwzuzn8lxcpzemhxue69uhhyetvv9ujumn0wd68ytnzv9hxgq3q6vzjeglr653mrmyqvu0trwaq29az753wr9th3hyrm5p63kz2zu8qf0c27j). These tools will
```

be crucial to improving the Hashpool user experience and tooling after the
core protocol is finalized.

A related project called
[Axepool](https://hashpool.dev/dev-logs/dev-log-2025-04-05/#axepool) applies the same design to small-scale miners,
particularly those running [Bitaxe](/blog/bitcoin-grants-feb-2024#the-bitaxe) hardware. By issuing
[eHash](https://hashpool.dev/articles/what-is-hashpool/#ehash) (ecash tokens
backed by proof of work) directly for contributed hashpower, Axepool removes
the need for miners to run a hosted service and enables direct ecash payouts, making it
easier for everyday miners to receive frequent payouts. The Axepool design is
still in the conceptual phase and undergoing rapid iteration. Together, Hashpool
and Axepool demonstrate how ecash can replace traditional pool accounting
with direct, accountless mining rewards that are private and transferable.

> OpenSats, developments have focused on integrating ecash issuance directly into
> the [Stratum V2](https://stratumprotocol.org/) protocol. The connection between
> pool and mint was rebuilt to use [asynchronous
> channels](https://github.com/vnprc/hashpool/commit/44f24d5), making payout flows
> more reliable. New [message
> formats](https://github.com/vnprc/hashpool/commit/5888558) were added in line
> with Cashu standards, and the development environment was upgraded with [Nix
> Flakes](https://nix.dev/concepts/flakes.html), Bitcoin
> [regtest](https://developer.bitcoin.org/examples/testing.html) setups, and
> [command-line utilities](https://github.com/vnprc/hashpool/commit/3c232ce) for
> testing wallet balances and redemption. Together, these changes make Hashpool
> easier to test and operate under real-world mining conditions.
> 
> Hashpool also explored how ecash payouts could be extended through Lightning and
> payment requests, with an [active pull
> request](https://github.com/vnprc/hashpool/pull/62) open for Lightning
> integration. These additions would allow miners to instantly redeem their tokens
> for bitcoin, thereby broadening the utility of ecash-based rewards.
> 
> A related project called
> [Axepool](https://hashpool.dev/dev-logs/dev-log-2025-04-05/#axepool) applies the
> same design to small-scale miners, particularly those running
> [Bitaxe](/blog/bitcoin-grants-feb-2024#the-bitaxe) hardware. By issuing
> [eHash](https://hashpool.dev/articles/what-is-hashpool/#ehash) (ecash tokens
> backed by proof of work) directly for contributed hashpower, Axepool removes
> account requirements and minimum withdrawal limits, making it easier for
> everyday miners to receive frequent payouts. Together, Hashpool and Axepool
> demonstrate how ecash can replace traditional pool accounting with direct,
> accountless mining rewards that are private and transferable.

Author: ```suggestion

```
